subsys/bluetooth/services/ras/rreq/ras_rreq.c

```diff
@@ -0,0 +1,723 @@
+	rreq->rd_ready.subscribe_params.notify = ranging_data_ready_notify_func;
+	rreq->rd_ready.subscribe_params.value = BT_GATT_CCC_NOTIFY | BT_GATT_CCC_INDICATE;
+	rreq->rd_ready.subscribe_params.subscribe = subscribed_func;
+
+	return 0;
+}
+
+static int rd_overwritten_subscribe_params_populate(struct bt_gatt_dm *dm, struct bt_ras_rreq *rreq)
+{
+	const struct bt_gatt_dm_attr *gatt_chrc;
+	const struct bt_gatt_dm_attr *gatt_desc;
+
+	/* RAS Ranging Data Ready characteristic (mandatory) */
+	gatt_chrc = bt_gatt_dm_char_by_uuid(dm, BT_UUID_RAS_RD_OVERWRITTEN);
+	if (!gatt_chrc) {
+		return -EINVAL;
+	}
+
+	gatt_desc = bt_gatt_dm_desc_by_uuid(dm, gatt_chrc, BT_UUID_RAS_RD_OVERWRITTEN);
+	if (!gatt_desc) {
+		return -EINVAL;
+	}
+	rreq->rd_overwritten.subscribe_params.value_handle = gatt_desc->handle;
+
+	gatt_desc = bt_gatt_dm_desc_by_uuid(dm, gatt_chrc, BT_UUID_GATT_CCC);
+	if (!gatt_desc) {
+		return -EINVAL;
+	}
+	rreq->rd_overwritten.subscribe_params.ccc_handle = gatt_desc->handle;
+	rreq->rd_overwritten.subscribe_params.notify = ranging_data_overwritten_notify_func;
+	rreq->rd_overwritten.subscribe_params.value = BT_GATT_CCC_NOTIFY | BT_GATT_CCC_INDICATE;
+	rreq->rd_overwritten.subscribe_params.subscribe = subscribed_func;
+
+	return 0;
+}
+
+static int on_demand_ranging_data_subscribe(struct bt_ras_rreq *rreq)
+{
+	int err;
+
+	err = bt_gatt_subscribe(rreq->conn, &rreq->on_demand_rd.subscribe_params);
+	if (err) {
+		LOG_DBG("On-demand ranging data subscribe failed (err %d)", err);
+		return err;
+	}
+
+	LOG_DBG("On-demand ranging data subscribed");
+
+	return 0;
+}
+
+static int ranging_data_ready_subscribe(struct bt_ras_rreq *rreq, bt_ras_rreq_rd_ready_cb_t cb)
+{
+	int err;
+
+	err = bt_gatt_subscribe(rreq->conn, &rreq->rd_ready.subscribe_params);
+	if (err) {
+		LOG_DBG("Ranging data ready subscribe failed (err %d)", err);
+		return err;
+	}
+
+	LOG_DBG("Ranging data ready subscribed");
+	rreq->rd_ready.cb = cb;
+
+	return 0;
+}
+
+static int ranging_data_overwritten_subscribe(struct bt_ras_rreq *rreq,
+					      bt_ras_rreq_rd_overwritten_cb_t cb)
+{
+	int err;
+
+	err = bt_gatt_subscribe(rreq->conn, &rreq->rd_overwritten.subscribe_params);
+	if (err) {
+		LOG_DBG("Ranging data overwritten subscribe failed (err %d)", err);
+		return err;
+	}
+
+	LOG_DBG("Ranging data overwritten subscribed");
+	rreq->rd_overwritten.cb = cb;
+
+	return 0;
+}
+
+void bt_ras_rreq_free(struct bt_conn *conn)
+{
+	int err;
+	struct bt_conn_info info;
+	struct bt_ras_rreq *rreq = ras_rreq_find(conn);
+	if (!rreq) {
+		return 0;
+	}
+
+	LOG_DBG("Free rreq %p for conn %p", (void *)conn, (void *)rreq);
+
+	err = bt_conn_get_info(conn, &info);
+	if (err != 0) {
+		return err;
+	}
+
+	if (info.state == BT_CONN_STATE_CONNECTED) {
+		/* -EINVAL means already not subscribed */
+		err = bt_gatt_unsubscribe(conn, &rreq->cp.subscribe_params);
+		if (err != 0 && err != -EINVAL) {
+			LOG_DBG("Failed to unsubscribe to RAS-CP: %d", err);
+
+			return err;
```

Comment:
Any of these failing will leave a ref to the conn, maybe the errors should be ignored as the link is probably getting torn down?

Reply:
Good point